### ISSUE TITLE
Huge memory usage regression in dotnet 7 on linux

### ISSUE THREAD

**Issue by 8** [dotnet/runtime#79633]
### Description
A dotnet 7 application consumes much more memory on linux than a dotnet 6 or dotnet 5 application.

After upgrading our application to dotnet 7 we noticed a huge increase in memory consumption in linux, services that used 100 MB were now using twice that amount.

The base issue can be easily reproduced

1. create new console application 
2. make sure it does not exit instantly, e.g. by using a while loop like:
```cs
while (true)
{
    Console.WriteLine("Hello, World!");
    Thread.Sleep(TimeSpan.FromSeconds(1));
}
```
3. Verify that in your .csproj file you are targeting dotnet 7:
```xml
<TargetFramework>net7.0</TargetFramework>
```
4. Switch to a `Release` configuration
5. Start it under linux, e.g. by select `WSL` in the dropdown to launch it.
6. Verify the used memory, e.g. by running `ps aux` in your wsl shell and taking a look at the `RSS` column:
```
USER       PID %CPU %MEM    VSZ   RSS TTY      STAT START   TIME COMMAND
root      5335  2.5  0.4 273549912 103944 pts/6 Sl+ 17:17   0:00 /usr/bin/dotnet /mnt/c/projects/sources/DotNet7Test/DotNet7Test/bin/Release/net7.0/DotNet7Test.dll
```
7. Realize in horror that the console process consumes 100+ MB
8. Stop the debugging
9. Target dotnet 6 by updating your csproj file to:
```xml
<TargetFramework>net6.0</TargetFramework>
```
10. Start it under linux, e.g. by select `WSL` in the dropdown to launch it.
11. Verify the used memory, e.g. by running `ps aux` in your wsl shell and taking a look at the `RSS` column:
```
USER       PID %CPU %MEM    VSZ   RSS TTY      STAT START   TIME COMMAND
root      5383  1.0  0.0 2961956 25504 pts/6   Sl+  17:21   0:00 /usr/bin/dotnet /mnt/c/projects/sources/DotNet7Test/DotNet7Test/bin/Release/net6.0/DotNet7Test.dll
```
12. Wonder why the same process consumes only 25MB using dotnet 6.

### Configuration

Dotnet 7 on Linux x64, see above

### Regression?

Yes, this is a major performance regression in dotnet 7 compared to dotnet 5 or 6.

### Other information

Notes:
The memory consumption on windows on dotnet 7 is reasonable - it's just a little higher, but not by crazy amounts.
The difference between dotnet 5 and dotnet 6 on linux is negliable.

Take care,
Martin

**Comment by mangod9** [comment]
Hi @8, 
yeah this is something we are aware of for unconstrained memory environments. We are working on resolving it in a upcoming servicing release, a workaround is to set `DOTNET_GCHeapHardLimit` to something reasonable (based on expected app usage or physical memory). Its an absolute increase of `64mb` hence its exaggerated for hello world style apps. 

It also shouldnt be an issue for memory constrainted containers

**Comment by enclave-alistair** [comment]
Just to add a note that we deploy a single-file self-contained linux application to end-users machines, so cannot realistically set `DOTNET_GCHeapHardLimit`, and an unrequired increase of `64mb` will no doubt raise significant complaints.

**Comment by mangod9** [comment]
Ok, will update this issue once the fix is merged. Thanks

**Comment by EricStG** [comment]
We have a similar issue. 

We upgraded 2 services to .NET 7 / ASPNET Code 7, both at least doubled memory usage (one went from a max of 15% memory usage to 80%)
Both are running in container with a 512mb limit (in Cloud Run if that matters)

**Comment by mangod9** [comment]
this issue shouldn't be affecting constrained memory containers (in fact, our TechEmpower measurements show a reduction in working set). Would you be able to share more details and/or a dump to investigate the increase further?

**Comment by EricStG** [comment]
@mangod9 Absolutely, let me know what you need

**Comment by mangod9** [comment]
to start with if you could please collect GC traces using `dotnet-trace` for both 6 and 7 that would be helpful: https://learn.microsoft.com/en-us/dotnet/core/diagnostics/dotnet-trace. Here is a sample command line:
`dotnet trace collect -p <pid> -o <outputpath with .nettrace extension> --profile gc-collect --duration <in hh:mm:ss format>` 

Also want to ensure that the container indeed is running with 512mb (since with K8s if no limit is specified it might be assigned more)

**Comment by EricStG** [comment]
Unfortunately, we can't access the container in cloud run to run anything on a shell.
I'll try to see if I can use `dotnet trace` with `--command` and try to find a way to store the trace on a volume.

In the meantime, when I set `DOTNET_GCHeapHardLimit` to the same as the memory limit, memory usage is on part between .NET 6 and 7.

Is there a way for us to figure out what memory limit the runtime thinks it has? I'm suspecting that the information is not exposed properly on cloud run, or at least not the same way docker usually works.

**Comment by mangod9** [comment]
GC will use cgroups to determine memory limits as shown here: https://github.com/dotnet/runtime/blob/cb603fc4eb81043cc382b9f4c85e994dabe9f879/src/coreclr/gc/unix/cgroup.cpp#L370 

If setting HeapHardLimit is making things work then I would assume cloud run is not setting those appropriately. Hoping the heaphardlimit works for you till the fix is made.

**Comment by 8** [comment]
I am not sure I follow:

`DOTNET_GCHeapHardLimit` sounds to me like a hard limit - meaning that memory allocations will fail, once that limit is reached?

That would be a bigger issue of course, as I'd rather want to the process to work if it really requires that memory, but to not use (or free it) when not required as I cannot know in advance how much memory will be used when the load increases.

If so, I need to wait for the fix before pulling the trigger on the dotnet 7 upgrade.

Take care,
Martin

**Comment by mangod9** [comment]
well the container also has a limit of 512mb correct?

**Comment by 8** [comment]
I don't quite understand - could you explain what you mean?

To explain my scenario:
I am running 10+ different dotnet services on each bare metal linux server.
When I upgrade them to dotnet 7, then my servers RAM is exhausted because all services memory is increased drastically.
If I set a hard limit, if one instance of a service requires more RAM than the hard limit, it will fail, which is not an option.

Take care,
Martin

**Comment by ilia-cy** [comment]
Hello @mangod9, are there any updates regarding the fix?

**Comment by mangod9** [comment]
Hi @ilia-cy, this should be fixed as part of https://github.com/dotnet/runtime/pull/80173, so should be included in 7.0.3 which should be released next week. Thx.

**Comment by idan-cyber** [comment]
@mangod9 - Any updates on version 7.0.3? Thanks!

**Comment by richlander** [comment]
It should ship one month from today, 3/14.

I never realized that Pi day is one month after Valentines.

**Comment by 8** [comment]
Hi Rich,

super!
yes, I would - if you have a download link that would be great!

Take care,
Martin

**Comment by idan-cyber** [comment]
https://github.com/dotnet/core/blob/main/release-notes/7.0/7.0.3/7.0.3.md Thanks!

**Comment by idan-cyber** [comment]
https://versionsof.net/core/7.0/7.0.3/ - Thanks @richlander!

**Comment by declard** [comment]
The issue doesn't seem to be fixed in either 7.0.3 or 7.0.4
```
sh-4.2$ uname -a
Linux ip-********.aws 4.14.305-227.531.amzn2.x86_64 #1 SMP Tue Feb 14 09:55:28 UTC 2023x86_64 x86_64 x86_64 GNU/Linux
sh-4.2$ dotnet --list-runtimes
Microsoft.NETCore.App 6.0.1 [/usr/share/dotnet/shared/Microsoft.NETCore.App]
Microsoft.NETCore.App 6.0.2 [/usr/share/dotnet/shared/Microsoft.NETCore.App]
Microsoft.NETCore.App 6.0.4 [/usr/share/dotnet/shared/Microsoft.NETCore.App]
Microsoft.NETCore.App 6.0.8 [/usr/share/dotnet/shared/Microsoft.NETCore.App]
Microsoft.NETCore.App 6.0.9 [/usr/share/dotnet/shared/Microsoft.NETCore.App]
Microsoft.NETCore.App 6.0.10 [/usr/share/dotnet/shared/Microsoft.NETCore.App]
Microsoft.NETCore.App 6.0.11 [/usr/share/dotnet/shared/Microsoft.NETCore.App]
Microsoft.NETCore.App 6.0.13 [/usr/share/dotnet/shared/Microsoft.NETCore.App]
Microsoft.NETCore.App 6.0.14 [/usr/share/dotnet/shared/Microsoft.NETCore.App]
Microsoft.NETCore.App 6.0.15 [/usr/share/dotnet/shared/Microsoft.NETCore.App]
Microsoft.NETCore.App 7.0.0 [/usr/share/dotnet/shared/Microsoft.NETCore.App]
Microsoft.NETCore.App 7.0.2 [/usr/share/dotnet/shared/Microsoft.NETCore.App]
Microsoft.NETCore.App 7.0.3 [/usr/share/dotnet/shared/Microsoft.NETCore.App]
Microsoft.NETCore.App 7.0.4 [/usr/share/dotnet/shared/Microsoft.NETCore.App]
```

The following is a screenshot of the dotnet6 version of our app running on a machine:
![image](https://user-images.githubusercontent.com/23086456/225158383-7c542877-f63c-4870-b239-d87e89305f8b.png)

Compare it to the screenshot of the dotnet7 version of the same app running on the same machine and the same data set:
![image](https://user-images.githubusercontent.com/23086456/225158461-ffb19ba5-9ffe-4394-8df8-868d53ab90a7.png)

The executable is a generic host console app, it runs
```csharp
GCSettings.LargeObjectHeapCompactionMode = GCLargeObjectHeapCompactionMode.CompactOnce;
GC.Collect();
```
in both versions after finishing loading the data.

Heap analysis didn't show any object size/count difference.

Such twofold increase in memory is unacceptable when an app consumes a hundred gigs in prod environments.

**Comment by mangod9** [comment]
Hi @declard, your issue seems to be different than the original posted here. That was regarding a vanilla app consuming more memory. I suggest you create a new issue and provide some traces to compare 6 with 7. Would also be good to check if setting `DOTNET_GCName=libclrgc.so` has any effect on your scenario. Thx!

**Comment by declard** [comment]
@mangod9 thank you for the advice, it actually worked! The memory dropped back to dotnet6 size, so I guess the issue is in the new GC indeed.

I'll dig further to file a new issue when time allows

**Comment by 8** [comment]
After the update I get:

```bash
dotnet --version
7.0.202
```
After rebuilding and running the executable I can see the memory usage dropped from 100MB to 30MB:

```bash
root@N900D90FB:~# ps aux
USER       PID %CPU %MEM    VSZ   RSS TTY      STAT START   TIME COMMAND
root       157  0.3  0.1 273476000 30460 pts/1 Sl+  12:41   0:00 /usr/bin/dotnet /mnt/c/projects/sources/DotNet7Test/DotNet7Test/bin/Release/net7.0/DotNet7Test.dll
```

Thank you for the fix!

We can now complete the upgrade of our services to dotnet 7, thank you!

Take care,
Martin

**Comment by mangod9** [comment]
Thanks for validating @8 . @declard, yeah please open a separate issue so we could investigate your specific scenario.

**Comment by mangod9** [comment]
@mihaimyh , the config moves the GC mechanism back to .NET 6. I would suggest creating a separate issue since your issue would most likely be different that the original one here. If you need notice working set differences with this setting would be good to collect some traces so we could investigate further. Thx!
